### PR TITLE
[mir-opt] Handle return place in ConstProp and improve SimplifyLocals pass

### PR DESCRIPTION
**Pull request by wesleywiser**
Temporarily rebased on top of #66074. The top 2 commits are new.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/transform/const_prop.rs [hunk 1]
```diff
@@ -335,6 +348,16 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
     }
 
     fn get_const(&self, local: Local) -> Option<Const<'tcx>> {
+        if local == RETURN_PLACE {
+            // Try to read the return place as an immediate so that if it is representable as a
+            // scalar, we can handle it as such, but otherwise, just return the value as is.
+            if let Some(Ok(Ok(imm))) = self.ret.map(|ret| self.ecx.try_read_immediate(ret)) {
```
Comment (Centril) [inline src/librustc_mir/transform/const_prop.rs]: Consider `return match self.ret.map(...) {` instead here; seems simpler?

Comment (wesleywiser) [reply src/librustc_mir/transform/const_prop.rs]: As in:

```rust
return match self.ret.map(|ret| self.ecx.try_read_immediate(ret)) {
    Some(Ok(Ok(imm))) => Some(imm.into()),
    _ => self.ret,
};
```

**Comment by wesleywiser** [comment]
@bors try
@rust-timer queue

File: src/librustc_mir/transform/const_prop.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_mir/transform/const_prop.rs]: Yep :)

**Review by oli-obk — APPROVED** [review]

**Comment by wesleywiser** [comment]
As expected, this shows a slight improvement over #66074. I'll apply @Centril's suggestion, rebase out #66074, and then r+ per @oli-obk's approval review.

**Comment by mark-i-m** [comment]
Woah, look at that perf! :dash:

**Comment by wesleywiser** [comment]
Just a small note, the performance improvements are from #66074 which this PR used to be based on top of. We won't see those improvements until that PR is merged.

**Pull request merged**